skrub/_joiner.py

```diff
@@ -318,52 +316,47 @@ def transform(self, X, y=None):
 
         Parameters
         ----------
-        X : :obj:`~pandas.DataFrame`, shape [n_samples, n_features]
+        X : dataframe
             The main table, to be joined to the auxiliary ones.
         y : None
             Unused, only here for compatibility.
 
         Returns
         -------
-        :obj:`~pandas.DataFrame`
+        dataframe
             The final joined table.
         """
         del y
         check_is_fitted(self, "vectorizer_")
-        input_is_polars = is_polars(X)
-        X = self._check_dataframe(X)
+        X = self._main_check_input.transform(X)
         _join_utils.check_missing_columns(X, self._main_key, "'X' (the main table)")
         _join_utils.check_column_name_duplicates(
             X, self._aux_table, self.suffix, main_table_name="X"
         )
         main = self.vectorizer_.transform(
-            X[self._main_key].set_axis(self._aux_key, axis="columns")
+            ns.set_column_names(s.select(X, s.cols(*self._main_key)), self._aux_key)
+        )
+        aux_table = ns.reset_index(
+            _join_utils.add_column_name_suffix(self._aux_table, self.suffix)
         )
-        match_result = self._matching.match(main, self.max_dist_)
-        aux_table = _join_utils.add_column_name_suffix(
-            self._aux_table, self.suffix
-        ).reset_index(drop=True)
-        matching_col = match_result["index"].copy()
-        matching_col[~match_result["match_accepted"]] = -1
+        _match_result = self._matching.match(main, self.max_dist_)
+        matching_col = _match_result["index"].copy()
+        matching_col[~_match_result["match_accepted"]] = -1
         token = _utils.random_string()
         left_key_name = f"skrub_left_key_{token}"
         right_key_name = f"skrub_right_key_{token}"
-        left = X.assign(**{left_key_name: matching_col})
-        right = aux_table.assign(**{right_key_name: np.arange(aux_table.shape[0])})
-        join = pd.merge(
+        left = ns.with_columns(X, **{left_key_name: matching_col})
+        right = ns.with_columns(
+            aux_table, **{right_key_name: np.arange(ns.shape(aux_table)[0])}
+        )
+        join = _join_utils.left_join(
             left,
             right,
             left_on=left_key_name,
             right_on=right_key_name,
-            suffixes=("", ""),
-            how="left",
         )
-        join = join.drop([left_key_name, right_key_name], axis=1)
+        join = s.select(join, ~s.cols(left_key_name))
         if self.add_match_info:
             for info_key, info_col_name in self._match_info_key_renaming.items():
-                join[info_col_name] = match_result[info_key]
-        if input_is_polars:
-            import polars as pl
-
-            join = pl.from_pandas(join)
+                join = ns.with_columns(join, **{info_col_name: _match_result[info_key]})
```

Comment:
I think we should build the dictionary in the loop and call `with_columns` only once it will be faster and a bit clearer